src/main/java/org/reaktivity/nukleus/amqp/internal/stream/AmqpServerFactory.java

```diff
@@ -656,15 +714,254 @@ private int decodeFlow(
+            annotationRW.wrap(frameBuffer, 0, frameBuffer.capacity());
+        annotation.forEach(kv -> vv ->
+        {
+            BoundedOctetsFW value = vv.getAsAmqpBinary().get();
+            switch (kv.kind())
+            {
+            case SYMBOL1:
+                String symbolKey = kv.getAsAmqpSymbol().get().asString();
+                annotationBuilder.item(b -> b.key(k -> k.name(symbolKey))
+                    .value(vb -> vb.bytes(value.value(), 0, value.length())));
+                break;
+            case ULONG0:
+            case ULONG1:
+            case ULONG8:
+                long longKey = kv.getAsAmqpULong().get();
+                annotationBuilder.item(b -> b.key(k -> k.id(longKey))
+                    .value(vb -> vb.bytes(value.value(), 0, value.length())));
+                break;
+            }
+        });
+        return annotationBuilder.build();
+    }
+
+    private AmqpPropertiesFW decodeProperties(
+        AmqpMessagePropertiesFW property)
+    {
+        AmqpPropertiesFW.Builder propertyBuilder = propertyRW.wrap(frameBuffer, 0, frameBuffer.capacity());
+        if (property.hasMessageId())
+        {
+            propertyBuilder.messageId(b -> b.stringtype(property.messageId().asString()));
+        }
+        if (property.hasUserId())
+        {
+            propertyBuilder.userId(b -> b.bytes(property.userId().value(), 0, property.userId().length()));
+        }
+        if (property.hasTo())
+        {
+            propertyBuilder.to(property.to().asString());
+        }
+        if (property.hasSubject())
+        {
+            propertyBuilder.subject(property.subject().asString());
+        }
+        if (property.hasReplyTo())
+        {
+            propertyBuilder.replyTo(property.replyTo().asString());
+        }
+        if (property.hasCorrelationId())
+        {
+            propertyBuilder.correlationId(b -> b.stringtype(property.correlationId().asString()));
+        }
+        if (property.hasContentType())
+        {
+            propertyBuilder.contentType(property.contentType().asString());
+        }
+        if (property.hasContentEncoding())
+        {
+            propertyBuilder.contentEncoding(property.contentEncoding().asString());
+        }
+        if (property.hasAbsoluteExpiryTime())
+        {
+            propertyBuilder.absoluteExpiryTime(property.absoluteExpiryTime());
+        }
+        if (property.hasCreationTime())
+        {
+            propertyBuilder.creationTime(property.creationTime());
+        }
+        if (property.hasGroupId())
+        {
+            propertyBuilder.groupId(property.groupId().asString());
+        }
+        if (property.hasGroupSequence())
+        {
+            propertyBuilder.groupSequence((int) property.groupSequence());
+        }
+        if (property.hasReplyToGroupId())
+        {
+            propertyBuilder.replyToGroupId(property.replyToGroupId().asString());
+        }
+        return propertyBuilder.build();
```

Comment:
If we want to remove those calls we need to cast `property.replyToGroupId()` since the type is `StringFW` and the builder class's accessor takes `String8FW`

Reply:
I see, would also prefer to avoid the cast if we can. 😉 

Would it make sense to adjust the builder to take `StringFW` instead of `String8FW` and then trigger an `assert` if the length is too big?

That would seem to still work for `String8FW` but also permit generic `StringFW` and even sufficiently short `String16FW` and `String32FW` to work as well.

We can go with the cast for now, to get the UTF8 benefit, and take the builder enhancement as a secondary step.

Reply:
Yes it makes sense for the builder to take `StringFW` I think.

I took a note for future enhancement.

Reply:
👍